doc: Add `near-network` reference level documentation `Handshake` (#5621)
near/nearcore: docs/network.md

> # 4. Code flow - finding first peer to connect to.

Reviewer: Why does it start from 4?

Author: It starts at 1: https://github.com/near/nearcore/pull/5580

Each node maintains list of peers.

> New node will get initial list of peers from:

Reviewer: ```suggestion
New node will get an initial list of peers from:
```

> - Genesis config `network.boot_nodes`

Reviewer: Genesis config is not the right term here. This is in client config (config.json)

- Command line argument `--boot_nodes`

Peer to connect to is chosen at random from list of known nodes by `PeerManagerActor::sample_random_peer` method.

# 5. Edges & network - in code representation
`P2P network` is represented by list of `peers`, where each `peer` is represented by structure `PeerId`,
which is defined by `peer's` public key `PublicKey`.

> And list of edges, where each edge is represented by structure `Edge`.

Reviewer: ```suggestion
And a list of edges, where each edge is represented by the structure `Edge`.
```

`TcpStream::connect` to create new connection.

## 6.5 - Step 5 - Connection gets established
Once connection with outgoing peer gets established.
`try_connect_peer` method will be called.
And then new `PeerActor` will be created and started.
One `PeerActor` starts it will send `Handshake` message to outgoing node `B` over tcp connection.

This message contains `protocol_version`, node's `A` metadata, as well as all information necessary to create `Edge`.

```rust
pub struct Handshake {
    /// Current protocol version.
    pub(crate) protocol_version: u32,
    /// Oldest supported protocol version.
    pub(crate) oldest_supported_version: u32,
    /// Sender's peer id.
    pub(crate) sender_peer_id: PeerId,
    /// Receiver's peer id.
    pub(crate) target_peer_id: PeerId,
    /// Sender's listening addr.
    pub(crate) sender_listen_port: Option<u16>,
    /// Peer's chain information.
    pub(crate) sender_chain_info: PeerChainInfoV2,
    /// Represents new `edge`. Contains only `none` and `Signature` from the sender.
    pub(crate) partial_edge_info: PartialEdgeInfo,
}
```

## 6.6 - Step 6 - `Handshake` arrives at node `B`

> Node `B` receives `Handshake` message. Then performs various validation checks.

Reviewer: ```suggestion
Node `B` receives `Handshake` message. Then it performs various validation checks.
```

Reviewer: Could we briefly describe what the validation checks are?

If everything is successful, `PeerActor` will send `Consolidate` message to `PeerManagerActor`.
This message contains everything needed to add `PeerActor` to list of active connections in `PeerManagerActor`.

Otherwise, `PeerActor` will be stopped immediately or after some timeout.

> TODO: Rename `Consolidate` to `RegisterPeer` + document all attributes.

Reviewer: Could we link an issue here

Author: We already merged that PR. It's already renamed.

# 4. Code flow - finding first peer to connect to.
Each node maintains list of peers.

New node will get an initial list of peers from:
- Genesis config `network.boot_nodes`
- Command line argument `--boot_nodes`

> 

Reviewer 2: I'd also mention that after connecting to the first node, we'll get the routing table from it - and then we'll do additional connections (up to 40?) with random peers from that list.

Peer to connect to is chosen at random from list of known nodes by `PeerManagerActor::sample_random_peer` method.

# 5. Edges & network - in code representation
`P2P network` is represented by list of `peers`, where each `peer` is represented by structure `PeerId`,
which is defined by `peer's` public key `PublicKey`.
And a list of edges, where each edge is represented by the structure `Edge`.

Both are defined below.

# 5.1 PublicKey
We use two types of public keys:
- a 256 bit `ED25519` public key
- a 512 bit `Secp256K1` public key

Public keys are defined in `PublicKey` enum, which consists of those two variants.
```rust
pub struct ED25519PublicKey(pub [u8; 32]);
pub struct Secp256K1PublicKey([u8; 64]);
pub enum PublicKey {
    ED25519(ED25519PublicKey),
    SECP256K1(Secp256K1PublicKey),
}
```

# 5.2 PeerId
Each `peer` is uniquely defined by its `PublicKey`, and represented by `PeerId` struct.
```rust
pub struct PeerId(PublicKey);
```

# 5.3 Edge
Each `edge` is represented by `Edge` struct. It's documented below:

```rust

> pub struct Edge {

Reviewer 2: it will be hard to keep this documentation in sync with the code. Maybe here just mention what is inside, and point to the file where it is implemented.

Author: Ok, I rewrote this section.

When `PeerManagerActor` starts it starts to listen to a specific port.

## 6.1 - Step 1 - `monitor_peers_trigger` runs
`PeerManager` checks if we need to connect to another peer by running `PeerManager::is_outbound_bootstrap_needed` method.
If `true` we will try to connect to new node.
Let's call current node, node `A`.

## 6.2 - Step 2 - choosing node to connect to
Method `PeerManager::sample_random_peer` will be called, and it returns node `B` that we will try to connect to.

## 6.3 - Step 3 - `OutboundTcpConnect` message
`PeerManagerActor` will send to itself a message `OutboundTcpConnect` in order to connect to node `B`.

```rust
pub struct OutboundTcpConnect {
    /// Peer information of the outbound connection
    pub target_peer_info: PeerInfo,
}
```

## 6.4 - Step 4 - `OutboundTcpConnect` message
On receiving the message `handle_msg_outbound_tcp_connect` method will be called, which calls
`TcpStream::connect` to create new connection.

## 6.5 - Step 5 - Connection gets established
Once connection with outgoing peer gets established.
`try_connect_peer` method will be called.
And then new `PeerActor` will be created and started.

> One `PeerActor` starts it will send `Handshake` message to outgoing node `B` over tcp connection.

Reviewer 2: s/One/Once

This message contains `protocol_version`, node's `A` metadata, as well as all information necessary to create `Edge`.

```rust
pub struct Handshake {
    /// Current protocol version.
    pub(crate) protocol_version: u32,
    /// Oldest supported protocol version.
    pub(crate) oldest_supported_version: u32,
    /// Sender's peer id.
    pub(crate) sender_peer_id: PeerId,
    /// Receiver's peer id.
    pub(crate) target_peer_id: PeerId,
    /// Sender's listening addr.
    pub(crate) sender_listen_port: Option<u16>,
    /// Peer's chain information.
    pub(crate) sender_chain_info: PeerChainInfoV2,
    /// Represents new `edge`. Contains only `none` and `Signature` from the sender.
    pub(crate) partial_edge_info: PartialEdgeInfo,
}
```

## 6.6 - Step 6 - `Handshake` arrives at node `B`

Node `B` receives `Handshake` message. Then it performs various validation checks.
That includes:

> - Signature checks from both peers:

Reviewer 2: 'both peers' ? shouldn't handshake contain only signature from A?

Author: @mm-near Oh, sorry. I wrote the documentation assuming I'm talking about edge validation. I rewrote this part.

- Check that key is an ordered pair
- Filter out edges that are already known, by checking whenever another version of the edge exists.
- Other sanity checks.

If everything is successful, `PeerActor` will send `Consolidate` message to `PeerManagerActor`.
This message contains everything needed to add `PeerActor` to list of active connections in `PeerManagerActor`.

Otherwise, `PeerActor` will be stopped immediately or after some timeout.

```rust
pub struct RegisterPeer {
    pub(crate) actor: Addr<PeerActor>,
    pub(crate) peer_info: PeerInfo,
    pub(crate) peer_type: PeerType,
    pub(crate) chain_info: PeerChainInfoV2,
    // Edge information from this node.
    // If this is None it implies we are outbound connection, so we need to create our
    // EdgeInfo part and send it to the other peer.
    pub(crate) this_edge_info: Option<EdgeInfo>,
    // Edge information from other node.
    pub(crate) other_edge_info: EdgeInfo,
    // Protocol version of new peer. May be higher than ours.
    pub(crate) peer_protocol_version: ProtocolVersion,
}
```

> ## 6.7 - Step 7 - `PeerManagerActor` receives `Consolidate` message

Reviewer 2: There should also be some response from node B, right?

`PeerManagerActor` open tcp server, which listens to incoming connection.
It starts `RoutingTableActor`, which then starts `EdgeValidatorActor`.
When connection incoming connection gets accepted, it starts a new `PeerActor` on its own thread.

# 4. NetworkConfig

`near-network` reads configuration from `NetworkConfig`, which is a part `client config`.

Here is a list of features read from config
- `boot_nodes` - list of nodes to connect to on start
- `addr` - listening address
- `max_num_peers` - by default we connect up to 40 peers, current implementation supports upto 128 nodes.

# 5. Connecting to other peers.

> Each peer maintains list of known peers, it's store in the database.

Reviewer: ```suggestion
Each peer maintains list of known peers. They are stored in the database.
```
